command/sub8_missions/missions/buoy.py

```diff
@@ -2,132 +2,98 @@
-        yield sub.nh.sleep(0.5)
-
-    # Now we are 1m away from the buoy
-    print "BUOY MISSION - bumping!"
-    sub.move.forward(dist + .3).go(speed=SPEED)
-
-    print "BUOY MISSION - Bumped the buoy"
-    defer.returnValue(True)
-
+    def __init__(self, sub):
+        self.sub = sub
+        self.print_info = FprintFactory(title=MISSION).fprint
+        self.print_bad = FprintFactory(title=MISSION, msg_color="red").fprint
+        self.print_good = FprintFactory(title=MISSION, msg_color="green").fprint
+        self.found = {'red': None, 'green': None, 'yellow': None}
+        self.bumped = []
+        self.do_search = False
+        self.generate_pattern()
+
+    def generate_pattern(self):
+        self.moves = [[0, 0, 0.3],
+                      [0, 1.5, 0],
+                      [0, 0, -2*0.3],
+                      [0, -2*1.5, 0]]
+        self.move_index = 0
+        
+    @util.cancellableInlineCallbacks
+    def search(self):
+        while self.search:
+            info = 'FOUND: '
+            for color in self.found:
+                res = yield self.sub.vision_proxies.buoy.get_pose(target=color)
+                if res.found:
+                    self.found[color] = rosmsg_to_numpy(res.pose.pose.position)
+                if self.found[color] is not None:
+                    info += color + ' '
+                yield self.sub.nh.sleep(0.5) # Throttle service calls
+            self.print_info(info)
+
+    @util.cancellableInlineCallbacks
+    def pattern(self):
+        def err():
+            print_info('Search pattern canceled')
+        for i, move in enumerate(self.moves[self.move_index:]):
+            move = self.sub.move.relative(np.array(move)).go()
+            move.addErrback(err)
+            yield move
+            self.move_index = i + 1
+
+    @util.cancellableInlineCallbacks
+    def bump(self, buoy):
+        self.print_info("BUMPING {}".format(buoy))
+        yield self.sub.move.go() # Station hold
+        start_pose = self.sub.pose
+ 
+        buoy_position = self.found[buoy]
+        yield self.sub.move.depth(-buoy_position[2]).go()
+        yield self.sub.move.look_at_without_pitching(buoy_position).go()
+        dist = np.linalg.norm(buoy_position - self.sub.pose.position)
+        yield self.sub.move.set_position(buoy_position).forward(0.2).go()
+        self.print_good("{} BUMPED. Backing up".format(buoy))
+        yield self.sub.move.backward(3.5).go()
+
+    @util.cancellableInlineCallbacks
+    def run(self):
+        self.print_info("Enabling Perception")
+        self.sub.vision_proxies.buoy.start()
+
+        pattern = self.pattern()
+        self.do_search = True
+        search = self.search()
+        while len(self.bumped) != len(self.found):
+            for color in self.found:
+                if self.found[color] is not None and color not in self.bumped:
+                    pattern.cancel()
+                    yield self.bump(color)
```

Comment:
So right now it just bumps buoys as it finds them. This will actually not maximize points as order matters. I'll work on this.